Services/AzureManagementAdapter/AzureManagementAdapter.cs

```diff
@@ -0,0 +1,239 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Security;
+using System.Threading.Tasks;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Exceptions;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Http;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Runtime;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.AzureManagementAdapter
+{
+    public interface IAzureManagementAdapterClient
+    {
+        Task<MetricsResponseModel> PostAsync(MetricsRequestsModel requests);
+    }
+
+    public class AzureManagementAdapter : IAzureManagementAdapterClient
+    {
+        private const string DATE_FORMAT = "yyyy-MM-ddTHH:mm:ssZ";
+        private const string METRICS_API_VERSION = "2016-06-01";
+        private readonly IServicesConfig config;
+        private readonly IDeploymentConfig deploymentConfig;
+        private readonly IDiagnosticsLogger diagnosticsLogger;
+        private readonly IHttpClient httpClient;
+        private readonly ILogger log;
+        private SecureString secureAccessToken = new SecureString();
+        private DateTimeOffset tokenExpireTime = DateTimeOffset.UtcNow;
+
+        public AzureManagementAdapter(
+            IHttpClient httpClient,
+            IServicesConfig config,
+            IDeploymentConfig deploymentConfig,
+            ILogger logger,
+            IDiagnosticsLogger diagnosticsLogger)
+        {
+            this.httpClient = httpClient;
+            this.log = logger;
+            this.config = config;
+            this.deploymentConfig = deploymentConfig;
+            this.diagnosticsLogger = diagnosticsLogger;
+        }
+
+        /// <summary>
+        ///     Query Azure Managemet API for the IotHub metrics.
+        ///     More details in following docs:
+        ///     https://docs.microsoft.com/en-us/azure/monitoring-and-diagnostics/monitoring-rest-api-walkthrough
+        ///     https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-metrics
+        ///     https://docs.microsoft.com/en-us/rest/api/monitor/metrics/list
+        /// </summary>
+        /// <param name="token"></param>
+        /// <param name="requests"></param>
+        /// <returns></returns>
+        public async Task<MetricsResponseModel> PostAsync(MetricsRequestsModel requests)
+        {
+            if (this.AccessTokenIsNullOrEmpty()) await this.GetAadTokenAsync();
+
+            if (this.AccessTokenExpireSoon()) this.GetAadTokenAsync();
+
+            if (requests == null) requests = this.GetDefaultMetricsRequests();
```

Comment:
I recommend keeping the API for getting metrics separate from AzureManagementAdapter. In future we will add additional APIs/requests which will use azure mgmt and will not want to set metrics request as default.

Reply:
This `PostAsync` method is for fetching metrics only. Please consider build other methods for different senarios.